es-app/src/VirtualSystemManager.cpp

```diff
@@ -0,0 +1,415 @@
+}
+
+FileData::SortType VirtualSystemManager::getSortType(std::string desc) {
+	std::vector<FileData::SortType> SortTypes = FileSorts::SortTypes;
+	// find it
+	for(unsigned int i = 0; i < FileSorts::SortTypes.size(); i++)
+	{
+		const FileData::SortType& sort = FileSorts::SortTypes.at(i);
+		if(sort.description == desc)
+		{
+			return sort;
+		}
+	}
+	// if not found default to name, ascending
+	return FileSorts::SortTypes.at(0);
+}
+
+std::vector<std::string> VirtualSystemManager::stringToSystems(std::string systemString)
+{
+	// from a comma separated string, get a vector of system names
+	std::vector<std::string> strs;
+	boost::split(strs, systemString, boost::is_any_of(","));
+	return strs;
+}
+
+void VirtualSystemManager::loadVirtualSystems()
+{
+	loadAutoVirtualSystems();
+	// we will also load custom systems here in the future
+	loadCustomVirtualSystems();
+	// Now see which ones are enabled
+	loadEnabledListFromSettings();
+	// add to the main System Vector, and create Views as needed
+	updateSystemsList();
+}
+
+void VirtualSystemManager::updateSystemsList()
+{
+	// remove all Virtual Systems
+	for(auto sysIt = SystemData::sSystemVector.begin(); sysIt != SystemData::sSystemVector.end(); )
+	{
+		if ((*sysIt)->isVirtual())
+		{
+			sysIt = SystemData::sSystemVector.erase(sysIt);
+		}
+		else
+		{
+			sysIt++;
+		}
+	}
+
+	// add enabled ones
+	for(std::map<std::string, VirtualSystemData>::iterator it = mAllVirtualSystems.begin() ; it != mAllVirtualSystems.end() ; it++ )
+	{
+		if(it->second.isEnabled)
+		{
+			SystemData::sSystemVector.push_back(it->second.system);
+		}
+	}
+
+	// remove disabled gamelist views
+	// create gamelist views if needed
+	// iterate the map
+}
+
+void VirtualSystemManager::loadAutoVirtualSystems()
+{
+	for(std::map<std::string, VirtualSystemDecl>::iterator it = mVirtualSystemDecls.begin() ; it != mVirtualSystemDecls.end() ; it++ )
+	{
+		VirtualSystemDecl sysDecl = it->second;
+		if (!sysDecl.isCustom && !findVirtualSystem(sysDecl.name))
+		{
+			SystemData* newSys = new SystemData(sysDecl.name, sysDecl.longName, mVirtualEnvData, sysDecl.themeFolder, true);
+
+			FileData* rootFolder = newSys->getRootFolder();
+			FileFilterIndex* index = newSys->getIndex();
+			for(auto sysIt = SystemData::sSystemVector.begin(); sysIt != SystemData::sSystemVector.end(); sysIt++)
+			{
+				if ((*sysIt)->getName() != "retropie") {
+					std::vector<FileData*> files = (*sysIt)->getRootFolder()->getFilesRecursive(GAME);
+					for(auto gameIt = files.begin(); gameIt != files.end(); gameIt++)
+					{
+						bool include = (*gameIt)->getName() != "kodi";
+						switch(sysDecl.type) {
+							case AUTO_LAST_PLAYED:
+								include = include && (*gameIt)->metadata.get("playcount") > "0";
+								break;
+							case AUTO_FAVORITES:
+								include = include && (*gameIt)->metadata.get("favorite") == "true";
+								break;
+						}
+
+						if (include) {
+							FileData* newGame = new FileData(*gameIt, newSys);
```

Comment:
A non-scientific benchmark (few times tried, not enough to get a scientific average - also, it may depend on other resources). I just measured the amount of memory used and time to load on the Pi. Not ideal, but a reasonable, not-completely-accurate proxy for what to expect.

```
No Virtual Systems (so, what I expect it to be in the current build):
pre-loading-ES: 63MB used
post-loading-ES: 153MB

Collections - loading the games 100x to simulate a large library (5k games * 100 = 500k games loaded in new FileDatas)
pre-ES starting: 63MB memory used
post-ES-load: 538MB memory used (+385MB, give or take)
time taken to build systems: 79038ms (79s for 100 times / 500k games loaded)

Collections - loading 10x (5k games * 10 = 50k games loaded in new FileDatas)
pre-ES starting: 63MB memory used
post-ES-load: 189MB memory used (+35MB give or take)
time taken to build systems: 7908ms (7.9s for 10 times / 50k games loaded)

Collections - normal (5k extra games in FileDatas)
pre: 63MB used
post-load: 157MB used (+5MB give or take)
time taken to build systems: 792ms (0.79s for 1 time / 5k games)
```

With a library of 50k games we should be talking of an increase of 8secs in loading time, and about 35MB in RAM being used.

If loading 5k games, we're talking about <1sec added loading time, and about 5MB extra memory.

I'd vouch for keeping it as it is, given the architectural expectations between views, system data and file data, and the effort of refactoring that vs the actual resources being used here, but I'm always open to pushbacks.